sqldelight-compiler/src/main/kotlin/com/squareup/sqldelight/model/SqlStmt.kt

```diff
@@ -90,6 +92,78 @@ class SqlStmt private constructor(
     this.arguments = orderedArguments
   }
 
+  internal fun factoryProgramMethod(factoryClass: ClassName): MethodSpec {
+    val method = MethodSpec.methodBuilder(name)
+        .addModifiers(Modifier.PUBLIC)
+        .addParameter(ClassName.get("android.database.sqlite", "SQLiteProgram"), "program")
+
+    // The first arguments to the method will be any foreign factories needed.
+    arguments.map { it.argumentType.comparable }
+        .filterNotNull()
+        .distinctBy { it.tableInterface }
+        .filter { !it.isHandledType && it.tableInterface != null && it.tableInterface != factoryClass }
+        .forEach { method.addParameter(it.tableInterface!!.nestedClass("Factory"), it.factoryField()) }
+
+    arguments.forEach { argument ->
+      val parameter = ParameterSpec.builder(
+          argument.argumentType.comparable?.javaType ?: TypeName.OBJECT, argument.name)
+      if (argument.argumentType.comparable != null) {
+        parameter.addAnnotations(argument.argumentType.comparable.annotations())
+      }
+      method.addParameter(parameter.build())
+
+      var startedControlFlow = false
+      if (argument.argumentType.comparable == null || argument.argumentType.comparable.nullable) {
+        method.beginControlFlow("if (${argument.name} == null)")
+            .addStatement("program.bindNull(${argument.index})")
+        startedControlFlow = true
+      }
+      if (argument.argumentType.comparable == null) {
+        method.nextControlFlow("else if (${argument.name} instanceof String)")
+            .addStatement("program.bindString(${argument.index}, ${argument.name})")
+            .nextControlFlow(
+                "else if (${argument.name} instanceof \$T || ${argument.name} instanceof \$T)",
+                TypeName.FLOAT.box(),
+                TypeName.DOUBLE.box()
+            )
+            .addStatement("program.bindDouble(${argument.index}, (double) ${argument.name})")
+            .nextControlFlow(
+                "else if (${argument.name} instanceof \$T" +
+                " || ${argument.name} instanceof \$T" +
+                " || ${argument.name} instanceof \$T)",
+                TypeName.INT.box(),
+                TypeName.SHORT.box(),
+                TypeName.LONG.box()
+            )
+            .addStatement("program.bindLong(${argument.index}, (long) ${argument.name})")
+            .nextControlFlow("else if (${argument.name} instanceof \$T)", TypeName.BOOLEAN.box())
+            .addStatement("program.bindLong(${argument.index}, ${argument.name} ? 1 : 0")
+            .nextControlFlow("else if (${argument.name} instanceof \$T)", ArrayTypeName.BYTE)
```

Comment:
This should be `ArrayTypeName.of(TypeName.BYTE)`. Right now you're emitting `instanceof byte` which won't compile.